<?xml version="1.0"?>
<launch>
  <!-- Arguments -->
  <arg name="model" default="$(env TURTLEBOT3_MODEL)" doc="model type [burger, waffle, waffle_pi]"/>
  
  <!-- Centralized initial pose parameters -->
  <arg name="initial_pose_x" default="-2.0"/>
  <arg name="initial_pose_y" default="0.0"/>
  <arg name="initial_pose_z" default="0.0"/>
  <arg name="initial_pose_a" default="0.0"/> <!-- yaw in radians -->
  
  <!-- Other common arguments -->
  <arg name="open_rviz" default="true"/>
  <arg name="move_forward_only" default="false"/>
  <arg name="gui" default="true"/>
  <arg name="world_name" default="turtlebot3_world"/>
  <arg name="use_initial_pose_publisher" default="true" doc="Use dedicated node to publish initial pose"/>

  <!-- Launch the simulation with the centralized initial pose -->
  <include file="$(find turtlebot3-nav-noetic)/launch/turtlebot3_simulation.launch">
    <arg name="model" value="$(arg model)"/>
    <arg name="x_pos" value="$(arg initial_pose_x)"/>
    <arg name="y_pos" value="$(arg initial_pose_y)"/>
    <arg name="z_pos" value="$(arg initial_pose_z)"/>
    <arg name="yaw" value="$(arg initial_pose_a)"/>  <!-- Added yaw -->
    <arg name="gui" value="$(arg gui)"/>
    <arg name="world_name" value="$(arg world_name)"/>
  </include>

  <!-- Launch the navigation with the same centralized initial pose -->
  <include file="$(find turtlebot3-nav-noetic)/launch/turtlebot3_navigation.launch">
    <arg name="model" value="$(arg model)"/>
    <arg name="initial_pose_x" value="$(arg initial_pose_x)"/>
    <arg name="initial_pose_y" value="$(arg initial_pose_y)"/>
    <arg name="initial_pose_a" value="$(arg initial_pose_a)"/>
    <arg name="open_rviz" value="$(arg open_rviz)"/>
    <arg name="move_forward_only" value="$(arg move_forward_only)"/>
  </include>
  
  <!-- Launch the initial pose publisher if requested -->
  <group if="$(arg use_initial_pose_publisher)">
    <node pkg="turtlebot3-nav-noetic" type="initial_pose_publisher.py" name="initial_pose_publisher">
      <param name="initial_pose_x" value="$(arg initial_pose_x)"/>
      <param name="initial_pose_y" value="$(arg initial_pose_y)"/>
      <param name="initial_pose_z" value="$(arg initial_pose_z)"/>
      <param name="initial_pose_a" value="$(arg initial_pose_a)"/>
      <param name="publication_duration" value="5.0"/>
    </node>
  </group>
  
  <!-- Launch the obstacle avoidance node -->
  <node pkg="turtlebot3-nav-noetic" type="obstacle_avoidance.py" name="obstacle_avoidance" output="screen">
    <param name="use_move_base" value="true"/>
    <param name="goal_tolerance" value="0.3"/>
    <param name="obstacle_threshold" value="0.5"/>
    <param name="linear_speed" value="0.2"/>
    <param name="angular_speed" value="0.5"/>
  </node>
</launch>